server/filestore.go

```diff
@@ -7338,3 +7593,107 @@ func (ts *templateFileStore) Store(t *streamTemplate) error {
 func (ts *templateFileStore) Delete(t *streamTemplate) error {
 	return os.RemoveAll(filepath.Join(ts.dir, t.Name))
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Compression
+////////////////////////////////////////////////////////////////////////////////
+
+type CompressionInfo struct {
+	Algorithm    StoreCompression
+	OriginalSize uint64
+}
+
+func (c *CompressionInfo) MarshalMetadata() []byte {
+	b := make([]byte, 14) // 4 + potentially up to 10 for uint64
+	b[0], b[1], b[2] = 'c', 'm', 'p'
+	b[3] = byte(c.Algorithm)
+	n := binary.PutUvarint(b[4:], c.OriginalSize)
+	return b[:4+n]
+}
+
+func (c *CompressionInfo) UnmarshalMetadata(b []byte) (int, error) {
+	c.Algorithm = NoCompression
+	c.OriginalSize = 0
+	if len(b) < 5 { // 4 + min 1 for uvarint uint64
+		return 0, nil
+	}
+	if b[0] != 'c' || b[1] != 'm' || b[2] != 'p' {
+		return 0, nil
+	}
+	var n int
+	c.Algorithm = StoreCompression(b[3])
+	c.OriginalSize, n = binary.Uvarint(b[4:])
+	if n <= 0 {
+		return 0, fmt.Errorf("metadata incomplete")
+	}
+	return 4 + n, nil
+}
+
+func (alg StoreCompression) Compress(buf []byte) ([]byte, error) {
+	if len(buf) < checksumSize {
```

Comment:
Probably should be a separate const threshold.

Reply:
Since we need to preserve the checksum, the `checksumSize` seems like it makes sense, as we need to make sure the buffer is big enough to do that.

Reply:
What I mean is probably does not make sense if block is below 1-2MB or so.

Reply:
I wonder if that might be surprising behaviour, what threshold do you think makes most sense?

Reply:
1M? WDYT?

Reply:
Looking at the constants, maybe `defaultSmallBlockSize / 2` makes most sense, which would be 0.5MB today, but I’m also aware of the fact that some of the unit tests set the block size to be deliberately small and by adding in a threshold like this, we may lose out on test coverage of compression by accident.

Reply:
Make your best call and we can get this merged in tomorrow.

Reply:
I think the best thing to do for now is to leave it as-is so that the test coverage is maximal. If we think that compressing small blocks is hurting performance we can always follow it up later with another PR.